repo_remote.go

```diff
@@ -146,3 +146,179 @@ func RepoRemoveRemote(repoPath, name string, opts ...RemoveRemoteOptions) error
+}
+
+// RemoteGetURL retrieves URL(s) of a remote of the repository in given path.
+func (r *Repository) RemoteGetURL(name string, opts ...RemoteGetURLOptions) ([]string, error) {
+	return RepoRemoteGetURL(r.path, name, opts...)
+}
+
+// RemoteSetURLOptions contains arguments for setting an URL of a remote of the repository.
+// Docs: https://git-scm.com/docs/git-remote#Documentation/git-remote.txt-emget-urlem
+type RemoteSetURLOptions struct {
+	// Indicates whether to get push URLs instead of fetch URLs.
+	Push bool
+	// The timeout duration before giving up for each shell command execution.
+	// The default timeout duration will be used when not supplied.
+	Timeout time.Duration
+}
+
+// RepoRemoteSetURL sets first URL of the remote with given name of the repository in given path.
+func RepoRemoteSetURL(repoPath, name, newurl string, opts ...RemoteSetURLOptions) error {
+	var opt RemoteSetURLOptions
+	if len(opts) > 0 {
+		opt = opts[0]
+	}
+
+	cmd := NewCommand("remote", "set-url")
+	if opt.Push {
+		cmd.AddArgs("--push")
+	}
+
+	_, err := cmd.AddArgs(name, newurl).RunInDirWithTimeout(opt.Timeout, repoPath)
+	if err != nil && strings.Contains(err.Error(), "No such remote") {
+		return ErrRemoteNotExist
+	}
+	return err
+}
+
+// RemoteSetURL sets the first URL of the remote with given name of the repository.
+func (r *Repository) RemoteSetURL(name, newurl string, opts ...RemoteSetURLOptions) error {
+	return RepoRemoteSetURL(r.path, name, newurl, opts...)
+}
+
+// RepoRemoteSetURLRegex sets the first URL of the remote with given name and URL regex match of the repository in given path.
+func RepoRemoteSetURLRegex(repoPath, name, urlregex string, newurl string, opts ...RemoteSetURLOptions) error {
+	var opt RemoteSetURLOptions
+	if len(opts) > 0 {
+		opt = opts[0]
+	}
+
+	cmd := NewCommand("remote", "set-url")
+	if opt.Push {
+		cmd.AddArgs("--push")
+	}
+
+	_, err := cmd.AddArgs(name, newurl, urlregex).RunInDirWithTimeout(opt.Timeout, repoPath)
+	if err != nil {
+		if strings.Contains(err.Error(), "No such URL found") {
+			return ErrURLNotExist
+		}
+		if strings.Contains(err.Error(), "No such remote") {
+			return ErrRemoteNotExist
+		}
+		return err
+	}
+	return nil
+}
+
+// RemoteSetURLRegex sets the first URL of the remote with given name and URL regex match of the repository.
+func (r *Repository) RemoteSetURLRegex(name, urlregex, newurl string, opts ...RemoteSetURLOptions) error {
+	return RepoRemoteSetURLRegex(r.path, name, urlregex, newurl, opts...)
+}
+
+// RepoRemoteURLAdd adds an URL to the remote with given name of the repository in given path.
+func RepoRemoteURLAdd(repoPath, name, newurl string, opts ...RemoteSetURLOptions) error {
+	var opt RemoteSetURLOptions
+	if len(opts) > 0 {
+		opt = opts[0]
+	}
+
+	cmd := NewCommand("remote", "set-url", "--add")
+	if opt.Push {
+		cmd.AddArgs("--push")
+	}
+
+	_, err := cmd.AddArgs(name, newurl).RunInDirWithTimeout(opt.Timeout, repoPath)
+	return err
+}
```

Comment:
Any reason `--add` is not a field (`Add bool` )in `RemoteSetURLOptions` but an independent function? I'm going to make a push to merge these two functions.

Reply:
The logic behind here was that it is commonly used, and useful to separate it, but actually isn't on second thought, good catch.

Reply:
(they are separated in manpages asw)

![image](https://user-images.githubusercontent.com/38327267/155691194-d2affda3-2f0b-4e0e-a833-30a83c64ad03.png)

Reply:
Add does not take any regex.

Reply:
Generics might work here: allowing two different RemoteSetURLOptions.

Runtime checking or even git failing doesn't seem ok for me.

Reply:
I see, I missed the part it does not accept the regex. Agree to have its own function then 😅